ansible/devutils

```diff
@@ -82,6 +129,87 @@ def action_console(parameters):
                                     console_password=vars['creds']['console_password'][vars['console_type']])
         console_host.posix_shell()
 
+
+def pdu_action_on_dut(host, attrs, action):
+    ret = { 'DUT' : host, 'PDU status' : [], 'Summary' : [], 'Action' : action }
+    pdu_name = attrs['pdu_host'] if 'pdu_host' in attrs else None
+    if not pdu_name:
+        ret['Summary'].append('DUT has no PDU configuration')
+        return ret
+
+    pdu_list = get_pdu_info(pdu_name)
+    if not pdu_list:
+        ret['Summary'].append('PDU not found in inventory')
+        return ret
+
+    for pdu_info in pdu_list:
+        pdu_host = pdu_info['ansible_host'] if pdu_info and 'ansible_host' in pdu_info else None
+        p_name = pdu_info['inventory_hostname'] if pdu_info and 'inventory_hostname' in pdu_info else None
+        if not pdu_host or not p_name:
+            ret['Summary'].append('No PDU IP or name')
+            continue
+
+        rec = { 'PDU' : p_name, 'PDU_IP' : pdu_host, 'status' : [] }
+        ret['PDU status'].append(rec)
+
+        controller = get_psu_controller(pdu_host, host, pdu_info)
+
+        if not controller:
+            ret['Summary'].append('Failed to communicate with controller {}'.format(p_name))
+            continue
+
+        rec['status'] = controller.get_psu_status()
+        if action == 'off':
+            for outlet in status:
+                controller.turn_off_psu(outlet['psu_id'])
+            rec['status'] = controller.get_psu_status()
+        elif action == 'on':
+            for outlet in status:
+                controller.turn_on_psu(outlet['psu_id'])
+            rec['status'] = controller.get_psu_status()
+        elif action != 'status':
+            ret['Summary'].append('Unsupported action {}.'.format(action))
+            continue
+
+    return ret
+
+def action_pdu(parameters, action):
+    hosts = parameters['hosts']
+    data = []
+    header = [ 'DUT', 'Action', 'PDU status', 'Summary' ]
+    for host, attrs in hosts.items():
+        g_task_runner.submit_task(host, pdu_action_on_dut, host=host, attrs=attrs, action=action)
+
+    for _, ret in g_task_runner.task_results():
+        status = ret['result']
+        if parameters['json']:
+            data.append(status)
+        else:
+            data.append([ status[x] for x in header ])
+
+    if parameters['json']:
+        print(json.dumps(data, indent=4))
+    else:
+        print(tabulate(data, headers=header, tablefmt='grid'))
+
+
+def action_pdu_status(parameters):
+    action_pdu(parameters, 'status')
+
+
+def action_pdu_off(parameters):
+    action_pdu(parameters, 'off')
+
+
+def action_pdu_on(parameters):
+    action_pdu(parameters, 'on')
+
+
+def action_pdu_reboot(parameters):
+    action_pdu(parameters, 'off')
+    action_pdu(parameters, 'on')
```

Comment:
I think it would be better to add a delay between ```power down``` and ```power up``` to ensure a safe power cycle.

Reply:
Another minor suggestion, the ```pdu_reboot``` will generate two outputs (one for power on and one for power off). I think it's better to supress the first output.

Reply:
Sleep is not practically needed but I added a one second sleep anyways.

I kept the off/on output but merged them in a same table to show the status change.